core/src/main/java/org/apache/calcite/rel/rules/BottomDecorrelateRule.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Correlate;
+import org.apache.calcite.sql2rel.RelDecorrelator;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This rule invokes the standard decorrelator {@link RelDecorrelator} on subplans of a query
+ * plan that are rooted at a {@link Correlate} node.  By invoking this rule repeatedly
+ * one essentially obtains a bottom-up application of the decorrelator, where
+ * {@link Correlate} nodes closer to leaves are processed first.  This strategy may
+ * enable decorrelating plans where applying the decorrelator at the root may fail.
+ * (The standard decorrelator is all-or-nothing, if it fails, it returns the original plan
+ * unchanged.  This strategy may still simplify some plans by decorrelating parts of them).
+ */
+@Value.Enclosing
+public class BottomDecorrelateRule
+    extends RelRule<BottomDecorrelateRule.Config>
+    implements TransformationRule {
+
+  protected BottomDecorrelateRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    Correlate cor = call.rel(0);
+    for (RelNode input : cor.stripped().getInputs()) {
+      if (containsCorrelate(input)) {
```

Comment:
Does this `input` need to call `stripRecursively`?

Reply:
that's how I wrote it originally, but it seems to work this way too.
Maybe I can call it inside containsCorrelate.